File: fixtures/javascript-upb05-configuration/controlled.js
/** @typedef {Object} ClockSample
 * @property {bigint} UnixMilliseconds
 * @property {bigint} Sequence
 */
export class ClockSample { constructor(UnixMilliseconds, Sequence) { this.UnixMilliseconds = UnixMilliseconds; this.Sequence = Sequence; } }

/** @param {ClockSample} sample @returns {bigint} */
export function ObserveClock(sample) { return sample.UnixMilliseconds; }

/** @typedef {Object} RandomState
 * @property {bigint} Value
 */
export class RandomState { constructor(Value) { this.Value = Value; } }

/** @typedef {Object} Draw
 * @property {bigint} State
 * @property {bigint} Value
 */
export class Draw { constructor(State, Value) { this.State = State; this.Value = Value; } }

/** @param {RandomState} state @returns {Draw} */
export function Next(state) {
  const value = BigInt.asIntN(64, state.Value * 48271n + 1n);
  return new Draw(value, value);
}

/** @typedef {Object} ControlledCommand
 * @property {bigint} Delta
 */
export class ControlledCommand { constructor(Delta) { this.Delta = Delta; } }

/** @typedef {Object} ControlledState
 * @property {bigint} Value
 */
export class ControlledState { constructor(Value) { this.Value = Value; } }

/** @typedef {Object} ControlledResult
 * @property {bigint} Value
 */
export class ControlledResult { constructor(Value) { this.Value = Value; } }

/** @param {ControlledState} state @param {ControlledCommand} command @returns {ControlledResult} */
export function DispatchControlled(state, command) {
  console.log(true);
  return new ControlledResult(BigInt.asIntN(64, state.Value + command.Delta));
}

/** @param {ControlledState} state @param {ControlledCommand} command @returns {ControlledResult} */
export function ReplayControlled(state, command) { return new ControlledResult(BigInt.asIntN(64, state.Value + command.Delta)); }

File: fixtures/lua-upb05-configuration/controlled.lua
local Seme = assert(_G.Seme, "Seme adapters are required")

---@class ClockSample
---@field UnixMilliseconds seme.i64
---@field Sequence seme.i64
---@class RandomState
---@field Value seme.i64
---@class Draw
---@field State seme.i64
---@field Value seme.i64
---@class ControlledCommand
---@field Delta seme.i64
---@class ControlledState
---@field Value seme.i64
---@class ControlledResult
---@field Value seme.i64

---@seme-id 8055555555555555555555555555550e
---@param sample ClockSample
---@return seme.i64
local function ObserveClock(sample)
  return Seme.field(sample, "UnixMilliseconds")
end

---@seme-id 8055555555555555555555555555550f
---@param state RandomState
---@return Draw
local function Next(state)
  local value = Seme.add(Seme.multiply(Seme.field(state, "Value"), Seme.i64_literal("48271")), Seme.i64_literal("1"))
  return Seme.record("Draw", { "State", "Value" }, { State = value, Value = value })
end

---@seme-id 80555555555555555555555555555510
---@param state ControlledState
---@param command ControlledCommand
---@return ControlledResult
local function DispatchControlled(state, command)
  Seme.observe(true)
  return Seme.record("ControlledResult", { "Value" }, { Value = Seme.add(Seme.field(state, "Value"), Seme.field(command, "Delta")) })
end

---@seme-id 80555555555555555555555555555511
---@param state ControlledState
---@param command ControlledCommand
---@return ControlledResult
local function ReplayControlled(state, command)
  return Seme.record("ControlledResult", { "Value" }, { Value = Seme.add(Seme.field(state, "Value"), Seme.field(command, "Delta")) })
end

return { ObserveClock = ObserveClock, Next = Next, DispatchControlled = DispatchControlled, ReplayControlled = ReplayControlled }
